qcow2reader_test.go

```diff
@@ -0,0 +1,125 @@
+package qcow2reader
+
+import (
+	"fmt"
+	"io"
+	"math/rand"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"testing"
+)
+
+const MiB = int64(1 << 20)
+const GiB = int64(1 << 30)
+
+func BenchmarkRead(b *testing.B) {
+	const size = 512 * MiB
+	base := filepath.Join(b.TempDir(), "image")
+	if err := createTestImage(base, size); err != nil {
+		b.Fatal(err)
+	}
+	b.Run("qcow2", func(b *testing.B) {
+		qcow2 := base + ".qocw2"
+		if err := convertToQcow2(base, qcow2, false); err != nil {
+			b.Fatal(err)
+		}
+		b.StopTimer()
+		b.ResetTimer()
+		for i := 0; i < b.N; i++ {
+			b.StartTimer()
+			n, err := readEntireImage(qcow2)
+			b.StopTimer()
+			if err != nil {
+				b.Fatal(err)
+			}
+			if n != size {
+				b.Fatalf("Expected %d bytes, read %d bytes", size, n)
+			}
+		}
+	})
+	b.Run("compressed qcow2", func(b *testing.B) {
+		compressedQcow2 := base + ".compressed.qcow2"
+		if err := convertToQcow2(base, compressedQcow2, true); err != nil {
+			b.Fatal(err)
+		}
+		b.StopTimer()
+		b.ResetTimer()
+		for i := 0; i < b.N; i++ {
+			b.StartTimer()
+			n, err := readEntireImage(compressedQcow2)
+			b.StopTimer()
+			if err != nil {
+				b.Fatal(err)
+			}
+			if n != size {
+				b.Fatalf("Expected %d bytes, read %d bytes", size, n)
+			}
+		}
+	})
+}
+
+// createTestImage creates a 50% allocated raw image with fake data that
+// compresses like real image data.
+func createTestImage(filename string, size int64) error {
+	const chunkSize = 4 * MiB
+	file, err := os.Create(filename)
+	if err != nil {
+		return err
+	}
+	if err := file.Truncate(size); err != nil {
+		return err
+	}
+	reader := &Generator{}
+	for offset := int64(0); offset < size; offset += 2 * chunkSize {
+		_, err := file.Seek(offset, io.SeekStart)
+		if err != nil {
+			return err
+		}
+		chunk := io.LimitReader(reader, chunkSize)
+		if n, err := io.Copy(file, chunk); err != nil {
+			return err
+		} else if n != chunkSize {
+			return fmt.Errorf("expected %d bytes, wrote %d bytes", chunkSize, n)
+		}
+	}
+	return file.Close()
+}
+
+// Generator generates data that compresses like a real image (30%).
```

Comment:
> 30%

This should be a variable in the struct?

Reply:
It is not easy to make this a variable. Without shuffling the image compresses to 0.005 of the raw size. With shuffling the entire buffer to 0.5, and 1/2 buffer to 0.25.